packages/grid/x-grid-modules/src/components/checkbox-renderer.tsx

```diff
@@ -17,7 +17,7 @@ export const HeaderCheckbox: React.FC<ColParams> = React.memo(({ api }) => {
       const isAllSelected =
         api.getAllRowIds().length === event.rows.length && event.rows.length > 0;
       const hasNoneSelected = event.rows.length === 0;
-      setChecked(isAllSelected || !hasNoneSelected);
+      setChecked(isAllSelected && !hasNoneSelected);
```

Comment:
Are you sure?
<img width="399" alt="Capture d’écran 2020-09-14 à 20 43 02" src="https://user-images.githubusercontent.com/3165635/93125099-eee47e00-f6ca-11ea-971a-51ec082c50e6.png">

vs.

<img width="372" alt="Capture d’écran 2020-09-14 à 20 42 56" src="https://user-images.githubusercontent.com/3165635/93125083-eb50f700-f6ca-11ea-9291-e02525dbe553.png">

Reply:
For me, checked is when all checkbox are checked and none are unchecked
But up to you, I can change it

Reply:
It makes me think of https://github.com/mui-org/material-ui/issues/20476. No specific preference. It could simply surface that the **intermediate state should always have the same visual look** no matter the checked state. At least it seems to be what bootstrap, [ant design](https://stackblitz.com/edit/react-hjpsz4?file=index.js).

@eps1lon any tips on the matter?

Reply:
What's the question here? If this is about color then the mixed state should have the same color as the checked state. At least that's [how our Checkbox behaves](https://material-ui.com/components/checkboxes/#basic-checkboxes) so if it behaves different here then there should be a rationale that doesn't apply to the general purpose use case.

Also note that the color doesn't really matter anyway because of SC 1.4.1 in WCAG 2.x. If there's a difference between checked+indeterminate and checked+determinate then it needs to be encoded in something other than color as well. Though ARIA does not support both of these states but rather a single `"mixed"` state.

<!--
While `checked` and `indeterminate` are different properties in the DOM and you can observe all 4 states ARIA can only have 3 states for `checked`: `true`, `false` and `'mixed'`. If I recall correctly then this issue is a real rabbit hole with a bunch of quirks across browsers and within React (related to the timing when you set `indeterminate` and how it affects the `:checked` pseudo class. 
-->

Reply:
OK awesome. I think that it perfectly answer the question, when `indeterminate` is set:

- It doesn't matter if checked is true or false, it's in a mixed state. 
- The color shouldn't change based on the checked state, if it does, it could incentive people to break 1.4.1. 

So as the next step, I propose:

- Here, we can do the change or note, it doesn't matter, so maybe use the new logic.
- On the main repository side, update the CSS so indeterminate unchecked has the primary color.

Reply:
> On the main repository side, we update the CSS so indeterminate unchecked has the primary color.

@dtassone Do you want to use this issue as an opportunity to make your first contribution to the core? :)